src/detect-metadata.c

```diff
@@ -28,23 +28,287 @@
+
+    DetectSetupParseRegexes(PARSE_REGEX, &parse_regex, &parse_regex_study);
+    DetectSetupParseRegexes(PARSE_TAG_REGEX, &parse_tag_regex, &parse_tag_regex_study);
 }
 
-static int DetectMetadataSetup (DetectEngineCtx *de_ctx, Signature *s, const char *rawstr)
+/**
+ *  \brief Free a Metadata object
+ */
+void DetectMetadataFree(DetectMetadata *mdata)
 {
+    SCEnter();
+
+    SCFree(mdata);
+
+    SCReturn;
+}
+
+/* djb2 string hashing */
+static uint32_t StringHashFunc(HashTable *ht, void *data, uint16_t datalen)
+{
+    uint32_t hash = 5381;
+    int c;
+
+    while ((c = *(char *)data++))
+        hash = ((hash << 5) + hash) + c; /* hash * 33 + c */
+
+    hash = hash % ht->array_size;
+
+    return hash;
+}
+
+static char StringHashCompareFunc(void *data1, uint16_t datalen1,
+                               void *data2, uint16_t datalen2)
+{
+    int len1 = strlen((char *)data1);
+    int len2 = strlen((char *)data2);
+
+    if (len1 == len2 && memcmp(data1, data2, len1) == 0) {
+        return 1;
+    }
+
     return 0;
 }
 
+static void StringHashFreeFunc(void *data)
+{
+    SCFree(data);
+}
+
+int DetectMetadataHashInit(DetectEngineCtx *de_ctx)
+{
+    if (! DetectEngineMustParseMetadata())
+        return 0;
+
+    de_ctx->metadata_table = HashTableInit(4096, StringHashFunc, StringHashCompareFunc, StringHashFreeFunc);
+    if (de_ctx->metadata_table == NULL)
+        return -1;
+    return 0;
+}
+
+void DetectMetadataHashFree(DetectEngineCtx *de_ctx)
+{
+    if (de_ctx->metadata_table)
+        HashTableFree(de_ctx->metadata_table);
+}
+
+static const char *DetectMedatataHashAdd(DetectEngineCtx *de_ctx, const char *string)
+{
+    const char * hstring = (char *)HashTableLookup(de_ctx->metadata_table, (void *)string, strlen(string));
+    if (hstring) {
+        return hstring;
+    }
+
+    const char *astring = SCStrdup(string);
+    if (astring == NULL) {
+        return NULL;
+    }
+
+    if (HashTableAdd(de_ctx->metadata_table, (void *)astring, strlen(astring)) == 0) {
+        return (char *)HashTableLookup(de_ctx->metadata_table, (void *)astring, strlen(astring));
+    } else {
+        SCFree((void *)astring);
+    }
+    return NULL;
+}
+
+static int DetectMetadataParse(DetectEngineCtx *de_ctx, Signature *s, const char *metadatastr)
+{
+#define MAX_SUBSTRINGS 30
+    int ret = 0, res = 0;
+    int ov[MAX_SUBSTRINGS];
+
+    ret = pcre_exec(parse_regex, parse_regex_study, metadatastr,
+                    strlen(metadatastr), 0, 0, ov, MAX_SUBSTRINGS);
+    if (ret < 2) {
+        /* Only warn user that the metadata are not parsed due
+         * to invalid format */
+        SCLogInfo("signature metadata not in key value format, ret %" PRId32
+                    ", string %s", ret, metadatastr);
+        return 0;
+    }
+
+    char *saveptr = NULL;
+    size_t metadatalen = strlen(metadatastr)+1;
+    char rawstr[metadatalen];
+    strlcpy(rawstr, metadatastr, metadatalen);
+    char * kv = strtok_r(rawstr, ",", &saveptr);
```

Comment:
kv is not checked here, which is then passed into strlen and pcre_exec. Due to the do { } while loop all next calls are checked
@regit